File: 9/9.hs
import System.Environment
import Data.List
import Data.Function
 
main = do
  file <- readFile "input.txt"
  --let file = "{}\n{{{}}}\n{{},{}}\n{{{},{},{{}}}}\n{{<ab>},{<ab>},{<ab>},{<ab>}}\n{{<!!>},{<!!>},{<!!>},{<!!>}}\n{{<a!>},{<a!>},{<a!>},{<ab>}}" 
  putStrLn $ show $ score 1 file
  putStrLn $ show $ score2 file

score d [] = 0
score d (x:xs) = case x of 
  '{' -> d + score (d+1) xs 
  '}' ->     score (d-1) xs
  '<' ->     garbage d xs
  _   ->       score d xs

  where garbage d [] = 0
        garbage d (x:y:zs) = case x of
          '>' ->   score d (y:zs) 
          '!' -> garbage d    zs
          _   -> garbage d (y:zs)

score2 [] =  0
score2 (x:xs) = case x of
  '<' -> garbage2 xs 
  _   ->   score2 xs

  where garbage2 [] = 0
        garbage2 (x:y:zs) = case x of
          '>' ->       score2 (y:zs) 
          '!' ->     garbage2    zs 
          _   -> 1 + garbage2 (y:zs) 

File: 2017/9/9.scala
// JAVA_OPTS="-Xmx4G -Xss4G" scala 9.scala

val filename = "input.txt"

val t = io.Source.fromFile(filename).mkString
//val t = "{}\n{{{}}}\n{{},{}}\n{{{},{},{{}}}}\n{{<ab>},{<ab>},{<ab>},{<ab>}}\n{{<!!>},{<!!>},{<!!>},{<!!>}}\n{{<a!>},{<a!>},{<a!>},{<ab>}}" 

def score(d: Int, s: String): Int = {
  if (s.isEmpty) return 0
  s.head match {
    case '{' => d + score(d+1, s.tail)  
    case '}' =>     score(d-1, s.tail) 
    case '<' =>     garbage(d, s.tail) 
    case  _  =>       score(d, s.tail) 
  }
}

def garbage(d: Int, s: String): Int = {
  if (s.isEmpty) return 0
  s.head match {
    case '>' =>   score(d, s.tail) 
    case '!' => garbage(d, s.tail.tail) 
    case  _  => garbage(d, s.tail) 
  }
}
 
println(score(1, t))

//////////////////////////////////////

def score2(s: String): Int = {
  if (s.isEmpty) return 0
  s.head match {
    case '<' => garbage2(s.tail) 
    case _ =>     score2(s.tail) 
  }
}

def garbage2(s: String): Int = {
  if (s.isEmpty) return 0
  s.head match {
    case '>' =>       score2(s.tail) 
    case '!' =>     garbage2(s.tail.tail) 
    case  _  => 1 + garbage2(s.tail) 
  }
}
 
println(score2(t))
